raiden-ts/src/services/epics.ts

```diff
@@ -903,3 +749,297 @@ export const msMonitorNewBPEpic = (
+  const invalidatedRecipients = new Set<Address>();
+
+  if (paths) {
+    for (const { path, fee } of paths) {
+      const cleanPath = getCleanPath(path, address);
+      const recipient = cleanPath[0];
+      let shouldSelectPath = false;
+      let reasonToNotSelect = '';
+
+      if (invalidatedRecipients.has(recipient)) continue;
+      if (filteredPaths.length === 0) {
+        shouldSelectPath = directTransferIsPossible(action, latest, recipient);
+      } else if (recipient !== filteredPaths[0].path[0]) {
+        reasonToNotSelect = 'path: already selected another recipient';
+      } else if (fee.gt(filteredPaths[0].fee)) {
+        reasonToNotSelect = 'path: already selected a smaller fee';
+      } else {
+        shouldSelectPath = true;
+      }
+
+      if (shouldSelectPath) {
+        filteredPaths.push({ path: cleanPath, fee });
+      } else {
+        log.warn('Invalidated received route. Reason:', reasonToNotSelect, 'Route:', cleanPath);
+        invalidatedRecipients.add(recipient);
+      }
+    }
+  }
+
+  return filteredPaths;
+}
+
+function getPsfInfo$(
+  pfsByAction: PFS | null | undefined,
+  pfsByConfig: string | Address | null,
+  deps: RaidenEpicDeps,
+): Observable<PFS> {
+  if (pfsByAction) return of(pfsByAction);
+  else if (pfsByConfig) return pfsInfo(pfsByConfig, deps);
+  else {
+    const { log, latest$ } = deps;
+    return latest$.pipe(
+      pluck('pfsList'), // get cached pfsList
+      first((pfsList) => pfsList.length > 0), // if needed, wait for list to be populated
+      mergeMap((pfsList) => pfsListInfo(pfsList, deps)), // fetch pfsInfo from whole list & sort it
+      tap((pfsInfos) => log.info('Auto-selecting best PFS from:', pfsInfos)),
+      pluck(0), // pop best ranked
+    );
+  }
+}
+
+function getIouForPfs(
+  pfs: PFS,
+  tokenNetwork: Address,
+  deps: RaidenEpicDeps,
+): Observable<{ pfs: PFS; iou: Signed<IOU> | undefined }> {
+  if (pfs.price.isZero()) {
+    return of({ pfs, iou: undefined });
+  } else {
+    return prepareNextIOU$(pfs, tokenNetwork, deps).pipe(map((iou) => ({ pfs, iou })));
+  }
+}
+
+function requestPfs$(
+  pfs: PFS,
+  iou: Signed<IOU> | undefined,
+  tokenNetwork: Address,
+  address: Address,
+  target: Address,
+  value: string,
+  httpTimeout: number,
+): Observable<{ pfsResponse: Response; responseText: string; iou: Signed<IOU> | undefined }> {
+  const body = losslessStringify({
+    from: address,
+    to: target,
+    value: value,
+    max_paths: 10,
+    iou: iou
+      ? {
+          ...iou,
+          amount: UInt(32).encode(iou.amount),
+          expiration_block: UInt(32).encode(iou.expiration_block),
+          chain_id: UInt(32).encode(iou.chain_id),
+        }
+      : undefined,
+  });
+
+  return fromFetch(`${pfs.url}/api/v1/${tokenNetwork}/paths`, {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body,
+  }).pipe(
+    timeout(httpTimeout),
+    map((pfsResponse) => ({ pfsResponse, iou })),
```

Comment:
Nitpick: no need for this map, if you just pass iou to use it below, it's already available in the context

Reply:
I could swear I took that from the old code-base. :laughing:

Reply:
:heavy_check_mark: